internal/pkg/encoding/fxamacker_cbor.go

```diff
@@ -0,0 +1,192 @@
+func (encoder *FxamackerCborEncoder) EncodeUint64(obj uint64) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt encodes a int.
+func (encoder *FxamackerCborEncoder) EncodeInt(obj int) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt8 encodes a int8.
+func (encoder *FxamackerCborEncoder) EncodeInt8(obj int8) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt16 encodes a int16.
+func (encoder *FxamackerCborEncoder) EncodeInt16(obj int16) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt32 encodes a int32.
+func (encoder *FxamackerCborEncoder) EncodeInt32(obj int32) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt64 encodes a int64.
+func (encoder *FxamackerCborEncoder) EncodeInt64(obj int64) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeBool encodes a bool.
+func (encoder *FxamackerCborEncoder) EncodeBool(obj bool) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeString encodes a string.
+func (encoder *FxamackerCborEncoder) EncodeString(obj string) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeArray encodes an array.
+func (encoder *FxamackerCborEncoder) EncodeArray(obj interface{}) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeMap encodes a map.
+func (encoder *FxamackerCborEncoder) EncodeMap(obj interface{}) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeStruct encodes a struct.
+func (encoder *FxamackerCborEncoder) EncodeStruct(obj interface{}) error {
+	// check for and handle special structs
+	switch v := obj.(type) {
+	case *hamt.Node:
+		return encoder.encodeCbor(e.NewEncHAMTNode(v))
+	case *cbg.Deferred:
+		return encoder.encodeCbor(e.CbgDeferred{Raw: v.Raw})
+	}
+	return encoder.encodeCbor(obj)
+}
+
+// Bytes returns the encoded bytes.
+func (encoder FxamackerCborEncoder) Bytes() []byte {
+	return encoder.b.Bytes()
+}
+
+func (encoder *FxamackerCborEncoder) encodeCbor(obj interface{}) error {
+	// get cbor encoded bytes
+	raw, err := cbor.Marshal(obj, cbor.EncOptions{})
+	if err != nil {
+		return err
+	}
+
+	// write to buffer
+	encoder.b.Write(raw)
+
+	return nil
+}
+
+//
+// FxamackerCborDecoder
+//
+
+// DecodeValue decodes an primitive value.
+func (decoder *FxamackerCborDecoder) DecodeValue(obj interface{}) error {
+	return decoder.decodeCbor(obj)
+}
+
+// DecodeArray decodes an array.
+func (decoder *FxamackerCborDecoder) DecodeArray(obj interface{}) error {
+	return decoder.decodeCbor(obj)
+}
+
+// DecodeMap encodes a map.
+func (decoder *FxamackerCborDecoder) DecodeMap(obj interface{}) error {
+	return decoder.decodeCbor(obj)
+}
+
+// DecodeStruct decodes a struct.
+func (decoder *FxamackerCborDecoder) DecodeStruct(obj interface{}) error {
+	// check for and handle special structs
+	switch v := obj.(type) {
+	case *hamt.Node:
+		var encHAMTNode e.HAMTNode
+		err := decoder.decodeCbor(&encHAMTNode)
+		if err != nil {
+			return err
+		}
+		encHAMTNode.FillIPLDHamtNode(v)
+		return nil
+	case *cbg.Deferred:
```

Comment:
where are we using cbg? and what for?

Reply:
We should talk about this plus plan to remove duplicated hamt types offline